### ISSUE TITLE
Using a Service Account results in failed binding

### ISSUE THREAD

**Issue by mattkoch614** [adldap/adLDAP#59]
Using the following configuration options with a Windows Service Account, binding fails.  This is because the account was specifically created to query Active Directory, and does not have the same attributes as a typical user.  

```
array(
    'account_suffix'=>  '@domain.com',
    'domain_controllers'=>  array('192.168.0.1'), 
    'ad_port' => 123,
    'base_dn'   =>  'DC=DOMAIN,DC=COM',
    'admin_username' =>  'CN=service_acct,OU=Service Accts,OU=Resources,DC=DOMAIN,DC=COM',
    'admin_password' =>  'password'',
    'use_ssl' => false,
);
```

Because of this, the following connection fails with a **Bind to Active Directory failed. Check the login credentials and/or server details. AD said: Invalid credentials** message. 

```
$this->ldapBind = @ldap_bind($this->ldapConnection, $this->adminUsername . $this->accountSuffix, $this->adminPassword);
```

 I believe this to be due to the `account_suffix` option.  If I use a different user to bind with such as my own Windows credentials, I am able to connect. Similarly, if I set `'account_suffix'=>  '',` I can bind with the service account, but it forces all of my users to log in with their fully qualified account (username@domain.com) since it's no longer explicitly set.

**Comment by noleafclover614** [comment]
A "hacky' solution was to modify ```adldap.php``` with the following, but I assume that is a better solution?

```
 if(strpos($this->adminUsername, 'OU=Service Accts'))
 {
   $this->ldapBind = @ldap_bind($this->ldapConnection, $this->adminUsername, $this->adminPassword);
  } else {
 $this->ldapBind = @ldap_bind($this->ldapConnection, $this->adminUsername . $this->accountSuffix, $this->adminPassword);
 }
```

**Comment by ChadSikorra** [comment]
Is that essentially just a user account without a UPN defined, or are you trying to do this with a Managed Service Account?

**Comment by noleafclover614** [comment]
Interesting question, and I'll admit I'm not really familiar with the differences between a Managed Service Account and a "regular" account.  That being said, the ```connect()``` method of the adLDAP package always tacks on the account suffix when it is making the connection - so we will always end up with '@domain.com' unless we leave the parameter blank.  Is that the way we should be connecting?

**Comment by ChadSikorra** [comment]
I think the order of a bind attempt should probably be...

* Attempt a bind with the unmodified username supplied.
* If bind fails, attempt a bind with the username in `username@domain.com` (UPN) form.
* If bind fails, attempt a bind with the username in `domain\username` form.
* If the bind still fails, then throw an exception.

That would cover the case where a user can enter a full DN, GUID, or display name, but also fall back to attempting a UPN or sAMAccountName based auth when the domain name is not explicitly defined in the username.

I think the only draw back to this approach would be that the last response code you get, after trying to bind several times in a row with subsequent failures, would probably not be very helpful in determining what actually went wrong.

Interesting side note...the actual RFC for LDAPv3 doesn't specify what form the username should be in for a bind. It just says that it should be "The name of the directory object that the client wishes to bind as" (http://www.ietf.org/rfc/rfc2251.txt - section 4.2). Which I guess accounts for all the interesting forms AD accepts for the bind username.

**Comment by noleafclover614** [comment]
@chadsikorra - I like your approach. It tries a few options before failing instead of assuming the account is always in account@domain.com form.

**Comment by ChadSikorra** [comment]
I actually tried to implement this (in a different library - not adLDAP). The only drawback I ran into was that multiple bind attempts with a valid username, but invalid password, will rack up bad password attempts and lead to a locked account faster. This might not be an issue with the main admin account used for the library, but in functions like `authenticate($username, $password)` it will lock out users twice as fast. 

I think the best that can be done is to detect either a GUID, DN, UPN or "domain\username" form and not modify those. The only type that gets left out is the case that you're probably hitting against here unfortunately. My guess is that the "name" (display name value) of your account is different from your sAMAccountName or userPrincipalName on the account. I think it's honestly a little silly that Microsoft even considers that a valid username for a bind attempt.

The only other possibility I can think of is just offering an option to adLDAP for whether or not to suffix usernames and let the developer make their own choice.
